batch/batch/inst_coll_config.py

```diff
@@ -108,21 +136,78 @@ async def async_init(self):
+        records = self.db.execute_and_fetchall('''
+SELECT * FROM resources;
+''')
+        self.resource_rates = {record['resource']: record['rate'] async for record in records}
+
+    def select_pool_from_cost(self, cores_mcpu, memory_bytes, storage_bytes):
+        assert self.resource_rates is not None
+
+        optimal_result = None
+        optimal_cost = None
+        for pool in self.name_pool_config.values():
+            result = pool.convert_requests_to_resources(cores_mcpu, memory_bytes, storage_bytes)
+            if result:
+                maybe_cores_mcpu, maybe_memory_bytes, maybe_storage_gib = result
+                maybe_cost = pool.cost_per_hour(self.resource_rates,
+                                                maybe_cores_mcpu,
+                                                maybe_memory_bytes,
+                                                maybe_storage_gib)
+                if optimal_cost is None or maybe_cost < optimal_cost:
+                    optimal_cost = maybe_cost
+                    optimal_result = (pool.name, maybe_cores_mcpu, maybe_memory_bytes, maybe_storage_gib, None)
+        return optimal_result
+
+    def select_pool_from_worker_type(self, worker_type, cores_mcpu, memory_bytes, storage_bytes):
         for pool in self.name_pool_config.values():
             if pool.worker_type == worker_type:
-                maybe_cores_mcpu = pool.resources_to_cores_mcpu(cores_mcpu, memory_bytes, storage_bytes)
-                if maybe_cores_mcpu is not None:
-                    return (pool.name, maybe_cores_mcpu)
+                result = pool.convert_requests_to_resources(cores_mcpu, memory_bytes, storage_bytes)
+                if result:
+                    actual_cores_mcpu, actual_memory_bytes, acutal_storage_gib = result
+                    return (pool.name, actual_cores_mcpu, actual_memory_bytes, acutal_storage_gib, None)
         return None
 
-    def select_job_private(self, machine_type, storage_bytes):  # pylint: disable=no-self-use
+    def select_job_private(self, machine_type, storage_bytes):
         return self.jpim_config.convert_requests_to_resources(machine_type, storage_bytes)
+
+    def select_inst_coll(self, machine_type, preemptible, worker_type, req_cores_mcpu,
+                         req_memory_bytes, req_storage_bytes):
+        if preemptible and machine_type is None:
+            if worker_type is not None:
+                result = self.select_pool_from_worker_type(
+                    worker_type=worker_type,
+                    cores_mcpu=req_cores_mcpu,
+                    memory_bytes=req_memory_bytes,
+                    storage_bytes=req_storage_bytes)
+            else:
+                result = self.select_pool_from_cost(
+                    cores_mcpu=req_cores_mcpu,
+                    memory_bytes=req_memory_bytes,
+                    storage_bytes=req_storage_bytes)
+            if result is not None:
```

Comment:
The implication of this line not being `return result` only is that jobs that require more than 16 cores will have private instances spun up for them. Do we want this behavior????